### PR TITLE
cranelift: properly splatting bytes in emit_small_memset

### PR DESCRIPTION
**Pull request by MaxGraey**
Previously byte splatting was incomplete and incorrect for I32 and I64 types. This fix implements more efficient and correct byte propagation for all `int_type`'s.

### CODE REVIEW

File: cranelift/frontend/src/frontend.rs [hunk 1]
```diff
@@ -768,9 +768,9 @@ impl<'a> FunctionBuilder<'a> {
 
             let ch = u64::from(ch);
             let raw_value = if int_type == types::I64 {
-                (ch << 32) | (ch << 16) | (ch << 8) | ch
+                ch * 0x0101010101010101_u64
             } else if int_type == types::I32 {
-                (ch << 16) | (ch << 8) | ch
+                ch * 0x01010101_u64
```
Comment (bjorn3) [inline cranelift/frontend/src/frontend.rs:773]: Could you keep using shift and bitor?

Comment (MaxGraey) [reply cranelift/frontend/src/frontend.rs:773]: Yes, sure!

Comment (MaxGraey) [reply cranelift/frontend/src/frontend.rs]: But keep in mind LLVM can't optimize such series of shifts:
https://godbolt.org/z/nadxYq

Comment (bjorn3) [reply cranelift/frontend/src/frontend.rs]: I assumed that shift and bitor would be faster than multiplying, but according to llvm-mca, multiplying is faster for 32bit and 64bit ints.

Comment (MaxGraey) [reply cranelift/frontend/src/frontend.rs]: > I assumed that shift and bitor would be faster than multiplying, but according to llvm-mca, multiplying is faster for 32bit and 64bit ints.

No. 64-bit multiply will be always faster in this case. Even for 32-bit platforms.

btw trunk branck of LLVM already perform this optimization but still suboptimal. See:
https://godbolt.org/z/o3jfWv

Comment (bjorn3) [reply cranelift/frontend/src/frontend.rs]: Ok. Could you please revert back to multiplying?

**Comment by MaxGraey** [comment]
Hi. Could someone review this?

**Review by cfallin — COMMENTED** [review]
Thanks for the PR!

Could you add a bit more description to the PR itself as to what this is fixing? I think I can *infer* that the old code was not properly splatting the byte across all bytes in the I32/I64. But the title just says "fix emit_small_memset" and the commit message just says "fix". For folks in the future it's helpful to have both a commit message that describes the issue, possibly with an example of what went wrong, and the fix; as well as a description in the PR itself about this. Thanks!

**Comment by MaxGraey** [comment]
Done! Is it enough for description? Btw you could edit my description to more informative manner. I'm totally ok with it

**Comment by cfallin** [comment]
Looks good, thanks! Would you mind amending the commit to have the same description as the commit message?

**Review by cfallin — APPROVED** [review]

**Pull request merged**